core-rust/state-manager/src/protocol/protocol_updates/protocol_update_definition.rs

```diff
@@ -0,0 +1,124 @@
+// This file contains the protocol update logic for specific protocol versions
+
+use crate::protocol::*;
+use radix_engine::transaction::CostingParameters;
+use radix_engine::types::*;
+
+use crate::transaction::*;
+use crate::LoggingConfig;
+use transaction::validation::{NotarizedTransactionValidator, ValidationConfig};
+
+/// A protocol update definition consists of two parts:
+/// 1) Updating the current (state computer) configuration ("transaction processing rules").
+///    This includes: transaction validation, execution configuration, etc
+/// 2) Executing arbitrary state updates against the current database state.
+///    While the abstraction is quite flexible, the only concrete implementation at the moment
+///    only modifies the state through committing system transactions (e.g. substate flash).
+pub trait ProtocolUpdateDefinition {
+    /// Additional (static) config which can be used to re-configure the updater.
+    type Overrides: ScryptoDecode;
+
+    /// Returns the new configuration that the state computer
+    /// should use after enacting the given protocol version.
+    fn state_computer_config(network_definition: &NetworkDefinition)
+        -> ProtocolStateComputerConfig;
+
+    fn create_updater(
+        new_protocol_version: &str,
+        network_definition: &NetworkDefinition,
+        overrides: Option<Self::Overrides>,
+    ) -> Box<dyn ProtocolUpdater>;
+}
+
+pub trait ConfigurableProtocolUpdateDefinition {
+    fn resolve_state_computer_config(
+        &self,
+        network_definition: &NetworkDefinition,
+    ) -> ProtocolStateComputerConfig;
+
+    /// Must return `Ok` if the `raw_config` is None or passed validate.
+    fn create_updater_with_raw_overrides(
+        &self,
+        new_protocol_version: &str,
+        network_definition: &NetworkDefinition,
+        raw_config: Option<&[u8]>,
+    ) -> Result<Box<dyn ProtocolUpdater>, DecodeError>;
+
+    fn validate_raw_overrides(&self, raw_config: &[u8]) -> Result<(), DecodeError>;
```

Comment:
(just a hunch)

shouldn't it be `fn validate(raw) -> ParsedValidOverrides`?
wouldn't it make `create_updater_with_raw_overrides()` infallible?

alternatively: maybe this method does not really belong in the trait?

Reply:
Well if the raw overrides are incorrect, we'd want to know at config time I think? Rather than parsing them incorrectly?

But maybe I'm not following you?

Reply:
Hmm I'm going to merge `ProtocolDefinitionResolver` into `ProtocolConfig` I think.

It just makes more sense I think having one place to validate everything.

Reply:
> we'd want to know at config time I think?

Yes,

I was just hoping that `Unwrap is allowed because we have already validated the raw config` can be avoided (by the power of type-safety, my favourite feature of all serious languages).
However, I played a bit with it right now, and I now think that either the `ConfigurableProtocolUpdateDefinition` is a bit artificial OR the `Overrides` associated type brings unneeded-flexibility (static `Vec<TxBatch>` instead?) at the price of excessive-complexity.

These aren't huge problems, but since this PR does some cool simplifications already, I only wanted to point out one more opportunity. Ofc `TODO: future refactor` is cool.

Reply:
I've tweaked it a little as you suggest (i.e. it's infallible, but has an explicit documented precondition that it will panic if the provided content is invalid)

A core issue here is that different protocol update definitions have different types for their config, and so we can't have a typed Overrides.

We could have a ValidatedRawOverrides, but this still wouldn't necessarily capture validated with respect to the particular protocol definition.

Reply:
The overrides will be used in quite a few places in future I think:
* If/when we migrate genesis to be a protocol update
* Any protocol updates which touch consensus manager constants; or otherwise might want to be configured differently in test environments
* The `custom-` protocol updates for testing